README.md
# Simple agnostic package for IO related (files, dir, etc..)

![Agnostic](https://img.shields.io/badge/php-agnostic-blue.svg?style=flat-square)
[![Latest Version on Packagist](https://img.shields.io/packagist/v/padosoft/io.svg?style=flat-square)](https://packagist.org/packages/padosoft/io)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/padosoft/io/master.svg?style=flat-square)](https://travis-ci.org/padosoft/io)
[![Quality Score](https://img.shields.io/scrutinizer/g/padosoft/io.svg?style=flat-square)](https://scrutinizer-ci.com/g/padosoft/io)
[![Total Downloads](https://img.shields.io/packagist/dt/padosoft/io.svg?style=flat-square)](https://packagist.org/packages/padosoft/io)
[![SensioLabsInsight](https://img.shields.io/sensiolabs/i/e5f0885a-912d-4bd8-a6c4-76127603d2b6.svg?style=flat-square)](https://insight.sensiolabs.com/projects/e5f0885a-912d-4bd8-a6c4-76127603d2b6)

This is a simple agnostic package with few or zero dipendencies. 
This is ideal to provides support and helpers for creating packeges without include many and many dipendencies.

##Requires
  
- php: >=7.0.0
  
## Installation

You can install the package via composer:
``` bash
$ composer require padosoft/io
```

## Usage

Then here's an example of how to use `FileHelper` class:

Add a use statement `use Padosoft\Io\FileHelper;`

```php
<?php

if(FileHelper::checkDirExistOrCreate('public/upload', '0755')){
    echo 'dir already exists or now created!';
}else{
    echo 'OOPS! dir not exists and unable to create it!';
}

//get all files in public/upload/
$arrFiles = FileHelper::findFiles('public/upload/*');
var_dump($arrFiles);

//get all jpeg in public/upload/
$arrFiles = FileHelper::findFiles('public/upload/*.jpg');
var_dump($arrFiles);

//check if the path ends with slash
if(DirHelper::endsWithSlash('public/upload')){
    echo 'add a final slash...';
    $path = DirHelper::addFinalSlash('public/upload');
    echo $path;
}else{
    echo 'path already finish with slash.';
}

//get all directories in public/upload/
$arrFolders = DirHelper::findDirs('public/upload/*');
var_dump($arrFolders);

//truncate log to last 2000 bytes without truncate line
echo 'before:'.PHP_EOL;
echo file_get_contents('log.csv');
if(LogHelper::ftruncatestart('log.csv', 2000)){
    echo 'after:'.PHP_EOL;
    echo file_get_contents('log.csv');
}

```

## List of functions

### DirHelper

- isDirSafe
- checkDirExistOrCreate
- addFinalSlash
- addFinalSlashToAllPaths
- endsWithSlash
- endsWithStar
- endsWith
- startsWithSlash
- startsWith
- findDirs
- delete
- removeFinalSlash
- removeFinalSlashToAllPaths
- removeStartSlash
- removeStartSlashToAllPaths
- copy
- isLocal
- isAbsoluteUnix
- isAbsoluteWindows
- isAbsoluteWindowsRoot
- isAbsolute
- isRelative
- join
- njoin
- canonicalize
- isDirEmpty : Check if a directory is empty in efficent way. Check hidden files too.
- isDotDir : Check if an antry is linux dot dir (i.e.: . or .. )
- isReadable : Check if a path is a dir and is readable.

### FileHelper

- arrMimeType
- getPathinfoPart
- getDirname
- getFilename
- getFilenameWithoutExtension
- getFilenameExtension
- hasExtension
- variable2Array
- changeExtension
- unlinkSafe
- fileExistsSafe
- findFiles
- filePutContentsSafe
- getMimeType
- getMimeTypeByFinfo
- getMimeTypeByMimeContentType
- isReadable : Check if a path is a file and is readable.
- filPerms : Gets file or dir permissions in octal form ('0777') or not ('644').

### LogHelper

- ftruncatestart

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email instead of using the issue tracker.

## Credits
- [Lorenzo Padovani](https://github.com/lopadova)
- [All Contributors](../../contributors)

## About Padosoft
Padosoft (https://www.padosoft.com) is a software house based in Florence, Italy. Specialized in E-commerce and web sites.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All Notable changes to `io` will be documented in this file

# 1.9.0 - 2022-04-04

### CHANGE
- Update composer.json


# 1.8.0 - 2016-09-24

### ADD
- FileHelper::filPerms() : Gets file or dir permissions in octal form ('0777') or not ('644').

# 1.7.0 - 2016-09-17

### ADD
- isDirEmpty
- isDotDir
- DirHelper::isReadable
- FileHelper::isReadable

# 1.6.1 - 2016-09-12

### FIX
- FIX bug in join.

## 1.6.0 - 2016-09-11

### ADD

### DirHelper
- isLocal
- isAbsoluteWindows
- isAbsoluteUnix
- isAbsolute
- isRelative
- join
- njoin
- removeStartSlash
- removeStartSlashToAllPaths
- canonicalize
- split

### FileHelper
- hasExtension
- changeExtension

## 1.5.0 - 2016-09-07

### ADD
- ADD LogHelper ftruncatestart method.

## 1.4.0 - 2016-08-30

- ADD copy dir with all contents.
- ADD removeFinalSlashToAllPaths.

## 1.3.1 - 2016-08-28

- ADD arr of MimeType.
- CHANGE method name to camelcase.

## 1.3.0 - 2016-08-28

- ADD getMimeType.

## 1.2.0 - 2016-08-25

- ADD support for PHP 7.1.
- ADD getPathinfoPart simple pathinfo wrapper and related helpers.

## 1.1.0 - 2016-08-24

- ADD removeFinalSlash and delete new methods.

## 1.0.0 - 2016-08-15

- Initial release


tests/DirHelperTest.php
<?php

namespace Padosoft\Io\Test;

use Padosoft\Io\DirHelper;
use PHPUnit\Framework\TestCase;
use SebastianBergmann\CodeCoverage\Node\Directory;

class DirHelperTest extends TestCase
{
    use \Padosoft\Io\Test\PathDataProvider;

    protected function setUp(): void
    {
        //init files and paths needed for tests.
        $this->initFileAndPath();
    }

    protected function tearDown(): void
    {
        //remove created path during test
        $this->removeCreatedPathDuringTest();
    }

    /**
     * @param $expected
     * @return bool
     */
    protected function expectedIsAnException($expected)
    {
        if (is_array($expected)) {
            return false;
        }

        return strpos($expected, 'Exception') !== false
        || strpos($expected, 'PHPUnit_Framework_') !== false
        || strpos($expected, 'TypeError') !== false;
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider pathProvider
     */
    public function isDirSafe($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::isDirSafe($path);
        } else {
            $this->assertEquals($expected, DirHelper::isDirSafe($path));
        }
    }

    /**
     * @test
     * @param $path
     * @param $modeMask
     * @param $expected
     * @dataProvider pathAndMaskProvider
     */
    public function checkDirExistOrCreate($path, $modeMask, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::checkDirExistOrCreate($path, $modeMask);
        } else {
            $this->assertEquals($expected, DirHelper::checkDirExistOrCreate($path, $modeMask));
            $this->assertEquals($expected, is_dir($path));
        }
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider pathForAddFinalSlashProvider
     */
    public function addFinalSlash($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::addFinalSlash($path);
        } else {
            $this->assertEquals($expected, DirHelper::addFinalSlash($path));
        }
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider pathForAddFinalSlashToAllPathProvider
     */
    public function addFinalSlashToAllPaths($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::addFinalSlashToAllPaths($path);
        } else {
            $this->assertEquals($expected, DirHelper::addFinalSlashToAllPaths($path));
        }
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider pathForEndsWithSlashProvider
     */
    public function endsWithSlash($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::endsWithSlash($path);
        } else {
            $this->assertEquals($expected, DirHelper::endsWithSlash($path));
        }
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider pathForEndsWithStarProvider
     */
    public function endsWithStar($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::endsWithStar($path);
        } else {
            $this->assertEquals($expected, DirHelper::endsWithStar($path));
        }
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider pathForEndsWithSlashProvider
     */
    public function endsWith($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::endsWith($path, '/');
        } else {
            $this->assertEquals($expected, DirHelper::endsWith($path, '/'));
        }
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider pathForStartsWithSlashProvider
     */
    public function startsWithSlash($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::startsWithSlash($path);
        } else {
            $this->assertEquals($expected, DirHelper::startsWithSlash($path));
        }
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider pathForStartsWithSlashProvider
     */
    public function startsWith($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::startsWith($path, '/');
        } else {
            $this->assertEquals($expected, DirHelper::startsWith($path, '/'));
        }
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider filesToFindDirsProvider
     */
    public function findDirs($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::findDirs($path);
        } else {
            $result = DirHelper::findDirs($path);
            $this->assertTrue(is_array($result));
            $this->assertIsArray($result);
            $this->assertCount(count($expected), $result);
        }
    }

    /**
     * @test
     */
    public function copy()
    {
        DirHelper::copy(__DIR__ . '/resources', __DIR__ . '/resourcesNew', [], function ($src, $dest) {
            echo "copy $src to $dest" . PHP_EOL;
        });
        $this->assertFileExists(__DIR__ . '/resourcesNew/');
        $this->assertFileExists(__DIR__ . '/resourcesNew/dummy.txt');
        $this->assertFileExists(__DIR__ . '/resourcesNew/dummy.csv');
        $this->assertFileExists(__DIR__ . '/resourcesNew/subdir/');
        $this->assertFileExists(__DIR__ . '/resourcesNew/subdir/dummy.txt');
        $this->assertFileExists(__DIR__ . '/resourcesNew/emptydir/');
        DirHelper::delete(__DIR__ . '/resourcesNew');

        DirHelper::copy(__DIR__ . '/resources', __DIR__ . '/resourcesNew', [__DIR__ . '/resources/subdir']);
        $this->assertFileExists(__DIR__ . '/resourcesNew/');
        $this->assertFileExists(__DIR__ . '/resourcesNew/dummy.txt');
        $this->assertFileExists(__DIR__ . '/resourcesNew/dummy.csv');
        //$this->assertFileNotExists(__DIR__.'/resourcesNew/subdir/dummy.txt');
        DirHelper::delete(__DIR__ . '/resourcesNew');
    }

    /**
     * @test
     * @param $path
     * @param $canonicalized
     * @dataProvider provideCanonicalizationTests
     */
    public function Canonicalize($path, $canonicalized)
    {
        $this->assertSame($canonicalized, DirHelper::canonicalize($path));
    }

    /**
     * @test
     * @param $path
     * @param $isAbsolute
     * @dataProvider provideIsAbsolutePathTests
     */
    public function isAbsolute($path, $isAbsolute)
    {
        $this->assertSame($isAbsolute, DirHelper::isAbsolute($path));
    }

    /**
     * @test
     * @param $path
     * @param $isAbsolute
     * @dataProvider provideIsAbsolutePathTests
     */
    public function isRelative($path, $isAbsolute)
    {
        $this->assertSame(!$isAbsolute, DirHelper::isRelative($path));
    }

    /**
     * @test
     * @param $path
     * @param $isAbsolute
     * @dataProvider provideIsAbsolutePathUnixTests
     */
    public function isAbsoluteUnix($path, $isAbsolute)
    {
        $this->assertSame($isAbsolute, DirHelper::isAbsoluteUnix($path));
    }

    /**
     * @test
     * @param $path
     * @param $isAbsolute
     * @dataProvider provideIsAbsolutePathWindowsTests
     */
    public function isAbsoluteWindows($path, $isAbsolute)
    {
        $this->assertSame($isAbsolute, DirHelper::isAbsoluteWindows($path));
    }

    /**
     * @test
     * @param $path
     * @param $isLocal $isAbsolute
     * @dataProvider provideIsLocalTests
     */
    public function isLocal($path, $isLocal)
    {
        $this->assertSame($isLocal, DirHelper::isLocal($path));
    }

    /**
     * @test
     * @param $paths
     * @param $expected
     * @dataProvider provideJoinArrayTests
     */
    public function JoinArray($paths, $expected)
    {
        $this->assertSame($expected, DirHelper::join($paths));
    }

    /**
     * @test
     * @param $path1
     * @param $path2
     * @param $expected
     * @dataProvider provideJoinTests
     */
    public function Join($path1, $path2, $expected)
    {
        $this->assertSame($expected, DirHelper::join($path1, $path2));
    }

    /**
     * @test
     * @param $path1
     * @param $path2
     * @param $expected
     * @dataProvider provideNjoinTests
     */
    public function Njoin($path1, $path2, $expected)
    {
        $this->assertSame($expected, DirHelper::njoin($path1, $path2));
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider provideisDirEmptyTests
     */
    public function isDirEmpty($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::isDirEmpty($path);
        } else {
            $this->assertSame($expected, DirHelper::isDirEmpty($path));
        }
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider provideisDotDirTests
     */
    public function isDotDir($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::isDotDir($path);
        } else {
            $this->assertSame($expected, DirHelper::isDotDir($path));
        }
    }

    /**
     * @test
     * @param $path
     * @param $expected
     * @dataProvider provideisReadableTests
     */
    public function isReadableTest($path, $expected)
    {
        if ($this->expectedIsAnException($expected)) {
            $this->expectException($expected);
            DirHelper::isReadable($path);
        } else {
            $this->assertSame($expected, DirHelper::isReadable($path));
        }
    }
}


tests/LogHelperTest.php
<?php

namespace Padosoft\Io\Test;

use Padosoft\Io\LogHelper;
use PHPUnit\Framework\TestCase;

class LogHelperTest extends TestCase
{
    use \Padosoft\Io\Test\PathDataProvider;

    protected function setUp(): void
    {
        //init files and paths needed for tests.
        $this->initFileAndPath();

        $dir = __DIR__ . '/resources/';
        if (!is_dir($dir)) {
            mkdir($dir, '0777', true);
        }

        $file = $dir . 'log.txt';
        if (!file_exists($file)) {
            $strLog = '';
            for ($i = 0; $i < 100; $i++) {
                $strLog .= $i . ';uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
            }
            file_put_contents($file, $strLog);
        }
    }

    protected function tearDown(): void
    {
        $dir = __DIR__ . '/resources/';
        $file = $dir . 'log.txt';
        if (file_exists($file)) {
            unlink($file);
        }

        //remove created path during test
        $this->removeCreatedPathDuringTest();
    }

    /**
     * @param $expected
     * @return bool
     */
    protected function expectedIsAnException($expected)
    {
        if (is_array($expected)) {
            return false;
        }

        return strpos($expected, 'Exception') !== false
        || strpos($expected, 'PHPUnit_Framework_') !== false
        || strpos($expected, 'TypeError') !== false;
    }

    /**
     * @test
     */
    public function ftruncatestartTest()
    {
        $dir = __DIR__ . '/resources/';
        $file = $dir . 'log.txt';

        $this->assertEquals(true, LogHelper::ftruncatestart($file, 2000));
        $content = file_get_contents($file);

        $truncatedfilecontents = '83;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;'.PHP_EOL;
        $truncatedfilecontents .= '84;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;'.PHP_EOL;
        $truncatedfilecontents .= '85;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;'.PHP_EOL;
        $truncatedfilecontents .= '86;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;'.PHP_EOL;
        $truncatedfilecontents .= '87;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '88;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '89;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '90;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '91;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '92;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '93;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '94;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '95;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '96;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '97;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '98;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;
        $truncatedfilecontents .= '99;uno;due;tre;quattro;cinque;sei;sette;otto;nove;dieci;' . PHP_EOL;

        $this->assertEquals($truncatedfilecontents, $content);
    }
}


src/LogHelper.php
<?php

namespace Padosoft\Io;

/**
 * Class LogHelper
 * @package Padosoft\Io
 */
class LogHelper
{
    /**
     * Truncates a file to a given length but keep the end.
     * Takes the filepointer, handle, and truncates the file to length, size.
     * It will not just cut it but search for a newline so you avoid corrupting your csv or logfiles.
     * @param string $filename
     * @param $maxfilesize in byte
     * @return bool
     * @see http://php.net/manual/en/function.ftruncate.php#103591
     */
    public static function ftruncatestart(string $filename, $maxfilesize) : bool
    {
        if (!FileHelper::fileExistsSafe($filename)) {
            return false;
        }
        $size = filesize($filename);
        if ($size === false || ($size < $maxfilesize * 1.0)) {
            return false;
        }
        $maxfilesize = $maxfilesize * 0.5; //we don't want to do it too often...
        $fh = fopen($filename, "r+");
        if ($fh === false) {
            return false;
        }
        ftell($fh);
        fseek($fh, -$maxfilesize, SEEK_END);
        $drop = fgets($fh);
        $offset = ftell($fh);
        for ($x = 0; $x < $maxfilesize; $x++) {
            fseek($fh, $x + $offset);
            $c = fgetc($fh);
            fseek($fh, $x);
            fwrite($fh, $c);
        }
        ftruncate($fh, $maxfilesize - strlen($drop));
        fclose($fh);

        return true;
    }
}


src/FileHelper.php
<?php

namespace Padosoft\Io;

/**
 * Helper Class FileHelper
 * @package Padosoft\Io
 */
class FileHelper
{
    public static $arrMimeType = array(

        'txt' => 'text/plain',
        'htm' => 'text/html',
        'html' => 'text/html',
        'php' => 'text/html',
        'css' => 'text/css',
        'js' => 'application/javascript',
        'json' => 'application/json',
        'xml' => 'application/xml',
        'swf' => 'application/x-shockwave-flash',
        'flv' => 'video/x-flv',

        // images
        'png' => 'image/png',
        'jpe' => 'image/jpeg',
        'jpeg' => 'image/jpeg',
        'jpg' => 'image/jpeg',
        'gif' => 'image/gif',
        'bmp' => 'image/bmp',
        'ico' => 'image/vnd.microsoft.icon',
        'tiff' => 'image/tiff',
        'tif' => 'image/tiff',
        'svg' => 'image/svg+xml',
        'svgz' => 'image/svg+xml',

        // archives
        'zip' => 'application/zip',
        'rar' => 'application/x-rar-compressed',
        'exe' => 'application/x-msdownload',
        'msi' => 'application/x-msdownload',
        'cab' => 'application/vnd.ms-cab-compressed',

        // audio/video
        'mp3' => 'audio/mpeg',
        'qt' => 'video/quicktime',
        'mov' => 'video/quicktime',

        // adobe
        'pdf' => 'application/pdf',
        'psd' => 'image/vnd.adobe.photoshop',
        'ai' => 'application/postscript',
        'eps' => 'application/postscript',
        'ps' => 'application/postscript',

        // ms office
        'doc' => 'application/msword',
        'rtf' => 'application/rtf',
        'xls' => 'application/vnd.ms-excel',
        'ppt' => 'application/vnd.ms-powerpoint',

        // open office
        'odt' => 'application/vnd.oasis.opendocument.text',
        'ods' => 'application/vnd.oasis.opendocument.spreadsheet',
    );

    /**
     * Simple pathinfo wrapper.
     * @param string $filePath
     * @param int $fileInfOoptions
     * @return string
     * @see http://php.net/manual/en/function.pathinfo.php
     */
    public static function getPathinfoPart(string $filePath, int $fileInfOoptions) : string
    {
        if ($filePath === null || $filePath == '' || is_dir($filePath) || DirHelper::endsWithSlash($filePath)) {
            return '';
        }

        $info = pathinfo($filePath, $fileInfOoptions);

        if ($info == '.' && $fileInfOoptions == PATHINFO_DIRNAME) {
            return '';
        }
        return ($info !== null && $info != '') ? $info : '';
    }

    /**
     * Return the file name of file (without path and without extension).
     * Return empty string if $filePath is null, empty or is a directory.
     * Ex.: /public/upload/pippo.txt return '/public/upload'
     * @param string $filePath
     * @return string
     */
    public static function getDirname(string $filePath) : string
    {
        return self::getPathinfoPart($filePath, PATHINFO_DIRNAME);
    }

    /**
     * Return the file name of file (without path and with extension).
     * Return empty string if $filePath is null, empty or is a directory.
     * Ex.: /public/upload/pippo.txt return 'pippo.txt'
     * @param string $filePath
     * @return string
     */
    public static function getFilename(string $filePath) : string
    {
        return self::getPathinfoPart($filePath, PATHINFO_BASENAME);
    }

    /**
     * Return the file name of file (without path and without extension).
     * Return empty string if $filePath is null, empty or is a directory.
     * Ex.: /public/upload/pippo.txt return 'pippo'
     * @param string $filePath
     * @return string
     */
    public static function getFilenameWithoutExtension(string $filePath) : string
    {
        return self::getPathinfoPart($filePath, PATHINFO_FILENAME);
    }

    /**
     * Return the file name of file (without path and without extension).
     * Return empty string if $filePath is null, empty or is a directory.
     * Ex.: /public/upload/pippo.txt return '.txt'
     * @param string $filePath
     * @param bool $ignoreCase if set to true return lowercase extension
     * (Requires mbstring extension for correct multi-byte character handling in extension)
     *
     * @return string
     */
    public static function getFilenameExtension(string $filePath, bool $ignoreCase = false) : string
    {
        $ext = self::getPathinfoPart($filePath, PATHINFO_EXTENSION);
        return $ignoreCase ? self::toLower($ext) : $ext;
    }

    /**
     * Returns whether the path has an extension.
     *
     * @param string $path The path string
     * @param string|array|null $extensions If null or not provided, checks if an
     *                                       extension exists, otherwise checks for
     *                                       the specified extension or array of extensions
     *                                       (with or without leading dot)
     * @param bool $ignoreCase Whether to ignore case-sensitivity
     *                                       (Requires mbstring extension for correct
     *                                       multi-byte character handling in extension)
     *
     * @return bool true if the path has an (or the specified) extension, otherwise false
     *
     * @see https://github.com/laradic/support/blob/master/src/Path.php
     */
    public static function hasExtension(string $path, $extensions = null, bool $ignoreCase = false) : bool
    {
        $actualExtension = self::getFilenameExtension($path, $ignoreCase);

        // Only check if path has any extension
        if (null === $extensions) {
            return !empty($actualExtension);
        }

        // Make an array of extensions
        $extensions = self::variable2Array($extensions);

        foreach ($extensions as $key => $extension) {
            if ($ignoreCase) {
                $extension = self::toLower($extension);
            }
            // remove leading '.' in extensions array
            $extensions[$key] = ltrim($extension, '.');
        }

        return in_array($actualExtension, $extensions);
    }

    /**
     * if $extensions is not an array, return [$extensions].
     * @param $extensions
     * @return array
     */
    protected static function variable2Array($extensions) : array
    {
        if (is_array($extensions)) {
            return $extensions;
        }
        return [$extensions];
    }

    /**
     * Changes the extension of a path string.
     *
     * @param string $path The path string with filename.ext to change
     * @param string $extension New extension (with or without leading dot)
     *
     * @return string The path string with new file extension
     *
     * @see https://github.com/laradic/support/blob/master/src/Path.php
     */
    public static function changeExtension($path, $extension)
    {
        if ('' === $path) {
            return '';
        }
        $actualExtension = self::getFilenameExtension($path);
        $extension = ltrim($extension, '.');
        // No extension for paths
        if ('/' == substr($path, -1)) {
            return $path;
        }
        // No actual extension in path
        if (empty($actualExtension)) {
            return $path . ('.' == substr($path, -1) ? '' : '.') . $extension;
        }
        return substr($path, 0, -strlen($actualExtension)) . $extension;
    }

    /**
     * unlink file if exists.
     * Return false if exists and unlink fails or if filePath is a dir.
     * @param string $filePath
     * @return bool
     */
    public static function unlinkSafe(string $filePath) : bool
    {
        if (!FileHelper::fileExistsSafe($filePath)) {
            return false;
        }

        if (DirHelper::isDirSafe($filePath)) {
            return false;
        }

        return unlink($filePath);
    }

    /**
     * Check if passed file exists or not.
     * If dir passed return false.
     * @param string $filePath
     * @return bool
     */
    public static function fileExistsSafe(string $filePath) : bool
    {
        if ($filePath == '') {
            return false;
        }

        if (DirHelper::isDirSafe($filePath)) {
            return false;
        }

        return file_exists($filePath);
    }

    /**
     * Find files matching a pattern (recursive with matched files in subdirs).
     * Returns an array containing the matched files (full path and not directories),
     * an empty array if no file matched or on error.
     * @param string $fileNamePattern if is null it set to base_path()/* if exists otherwise __DIR__/* . It support glob() string pattern.
     * @param int $flags glob() Valid flags
     * @return array of files (full path)
     */
    public static function findFiles(string $fileNamePattern, int $flags = 0)
    {
        $fallback = [];

        if (($fileNamePattern === null || $fileNamePattern == '') && function_exists('base_path')) {
            $fileNamePattern = DirHelper::addFinalSlash(base_path()) . '*';
        } elseif ($fileNamePattern === null || $fileNamePattern == '') {
            $fileNamePattern = __DIR__ . '/*';
        }

        if (DirHelper::endsWithSlash($fileNamePattern)) {
            $fileNamePattern .= '*';
        }

        $files = glob($fileNamePattern, $flags);

        //remove array of empty string
        $files = array_filter($files, function ($k) {
            return ($k !== null && $k != '');
        });

        if (empty($files)) {
            return $fallback;
        }

        foreach (glob(dirname($fileNamePattern) . '/*', GLOB_ONLYDIR | GLOB_NOSORT) as $dir) {

            if (empty($dir)) {
                continue;
            }

            $files = array_merge($files, self::findFiles($dir . '/' . basename($fileNamePattern), $flags));
            $files = array_filter($files, function ($k) {
                return ($k !== null && $k != '');
            });
        }

        $files = array_filter($files, function ($k) {
            return (!is_dir($k));
        });

        return $files === false ? $fallback : $files;
    }

    /**
     * Equals to file_put_contents but safe, i.e.
     * accept empty string and return false without raise an error,
     * accept a directory and return false without raise an error,
     * and if $forceCreateDirIfNotExists is set to true and path doesn't exists, file_put_contents fails
     * so, this class, try to create the complete path before save file.
     * @param string $filename file name including folder.
     * example :: /path/to/file/filename.ext or filename.ext
     * @param string $data The data to write.
     * @param bool $forceCreateDirIfNotExists if true and path not exists, try to create it.
     * @param string $modeMask The mask applied to dir if need to create some dir.
     * @param int $flags same flags used for file_put_contents.
     * @see more info: http://php.net/manual/en/function.file-put-contents.php
     * @return bool TRUE file created succesfully, return FALSE if failed to create file.
     */
    public static function filePutContentsSafe(
        string $filename,
        string $data,
        bool $forceCreateDirIfNotExists = true,
        string $modeMask = '0755',
        int $flags = 0
    ) : bool
    {
        if ($filename == '') {
            return false;
        }

        //check if a directory passed ($filename ends with slash)
        if (DirHelper::endsWithSlash($filename)) {
            return false;
        }

        $dirName = dirname($filename);

        if (!$forceCreateDirIfNotExists && !DirHelper::isDirSafe($dirName)) {
            return false;
        }

        if (!DirHelper::checkDirExistOrCreate(DirHelper::addFinalSlash($dirName), $modeMask)) {
            return false;
        }

        return file_put_contents($filename, $data, $flags);
    }

    /**
     * Return mime type of a passed file in optimized mode.
     * @param string $fullPathFile
     * @return string
     */
    public static function getMimeType(string $fullPathFile) : string
    {
        $mime_types = self::$arrMimeType;
        $ext = strtolower(self::getFilenameExtension($fullPathFile));
        if (array_key_exists($ext, $mime_types)) {
            return $mime_types[$ext];
        }
        $mimetype = self::getMimeTypeByMimeContentType($fullPathFile);
        if (isNotNullOrEmpty($mimetype)) {
            return $mimetype;
        }
        $mimetype = self::getMimeTypeByFinfo($fullPathFile);
        if (isNotNullOrEmpty($mimetype)) {
            return $mimetype;
        }
        return 'application/octet-stream';
    }

    /**
     * Return mime type of a passed file using finfo
     * @param string $fullPathFile
     * @return string return empty string if it fails.
     */
    public static function getMimeTypeByFinfo(string $fullPathFile) : string
    {
        if (!function_exists('finfo_open')) {
            return '';
        }
        $finfo = finfo_open(FILEINFO_MIME);
        $mimetype = finfo_file($finfo, $fullPathFile);
        finfo_close($finfo);
        if ($mimetype === false) {
            return '';
        }
        return $mimetype;
    }

    /**
     * Return mime type of a passed file using mime_content_type()
     * @param string $fullPathFile
     * @return string return empty string if it fails.
     */
    public static function getMimeTypeByMimeContentType(string $fullPathFile) : string
    {
        if (!function_exists('mime_content_type')) {
            return '';
        }
        return mime_content_type($fullPathFile);
    }

    /**
     * Converts string to lower-case (multi-byte safe if mbstring is installed).
     *
     * @param string $str The string
     *
     * @return string Lower case string
     * @see https://github.com/laradic/support/blob/master/src/Path.php
     */
    private static function toLower($str)
    {
        if (function_exists('mb_strtolower')) {
            return mb_strtolower($str, mb_detect_encoding($str));
        }
        return strtolower($str);
    }

    /**
     * Check if $path is a file and is readable.
     * Return false if you pass a dir.
     * @param string $path
     * @return bool
     */
    public static function isReadable(string $path):bool
    {
        return self::fileExistsSafe($path) && is_readable($path);
    }

    /**
     * Gets file or dir permissions in octal form ('0777') or not ('644').
     * @param string $file
     * @param bool $octal default false i.e. return '644'
     * @return bool|string return false if it fails.
     */
    public static function filPerms(string $file, bool $octal = false)
    {
        if (!file_exists($file)) {
            return false;
        }

        $perms = fileperms($file);
        if($perms===false){
            return false;
        }

        $cut = $octal ? -4 : -3;

        return substr(decoct($perms), $cut);
    }
}
